server/test/integration/com/thoughtworks/go/server/service/UserServiceIntegrationTest.java

```diff
@@ -609,6 +611,73 @@ public void shouldDisableAdminSelectionWhenUserIsMemberOfAdminRole() throws Exce
         assertThat(userService.getAdminAndRoleSelections(Arrays.asList("foo", "quux")).getAdminSelection(), is(new TriStateSelection(Admin.GO_SYSTEM_ADMIN, TriStateSelection.Action.add, false)));
     }
 
+    @Test
+    public void updateShouldUpdateEnabledState() throws Exception {
+        userDao.saveOrUpdate(new User("admin-user"));
+        userDao.saveOrUpdate(new User("user-1"));
+        userService.enable(Arrays.asList("user-1"), new HttpLocalizedOperationResult());
+
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        userService.update(new User("user-1"), TriState.FALSE, TriState.UNSET, null, null, result);
```

Comment:
Can I change the implementation to this -

```java
    public User update(User user, TriState enabled, TriState emailMe, String email, String checkinAliases, LocalizedOperationResult result) {
        if (enabled.isTrue()) {
            user.enable();
        }

        if (enabled.isFalse()) {
            user.disable();
        }

        if (email != null) {
            user.setEmail(email);
        }

        if (checkinAliases != null) {
            user.setMatcher(checkinAliases);
        }

        if (emailMe.isTrue()) {
            user.setEmailMe(true);
        }

        if (emailMe.isFalse()) {
            user.setEmailMe(false);
        }

        try {
            saveOrUpdate(user);
        } catch (ValidationException e) {
            result.badRequest(LocalizedMessage.string("USER_FIELD_VALIDATIONS_FAILED", e.getMessage()));
        }

        return user;
    }
```

Notice that it updates and mutates the same object and attempts to to save it in a single transaction.

This prevents a double load from the controller as well as from inside the service. The service does exactly what it advertises "update", no more.